### ISSUE TITLE
Store Lat/Long with location

### ISSUE THREAD

**Issue by stevejgordon** [HTBox/allReady#1570]
We use a location table for the locations of campaigns, events, itineraries etc.

We store the lat/long outside of this table currently. Would it make more sense to store this data with the location?

**Comment by mgmccarthy** [comment]
@stevejgordon, I'd rather leave the data where it is. Although it makes reads more difficult (extra `Includes`), having that data separate will help to alleviate contention around updates to any of the entities that rely on the lat long.

For example, when updating lat long (or doing a lookup via a service to get it), how often are we also changing the other data on the Campaign, Event, etc... entities? My guess is hardly at all.

**Comment by EmilyLuijbregts** [comment]
@stevejgordon - based upon the comment from @mgmccarthy - how would you like to proceed? Keep the task? Close? or go back to the product team?

**Comment by stevejgordon** [comment]
@mgmccarthy - I'm not 100% sure I understand your point, but it's early here :smile: - So we have the location which is essentially the address we already have to include when we present say a campaign details page. It feels like the lat/long are properties that location and once set, should not change (unless the whole address changes).

I'd need to dig into the use of locations elsewhere in the app though as I might be missing something more complex.

**Comment by mgmccarthy** [comment]
cc @EmilyLuijbregts 

LOL... @stevejgordon, I read the description wrong! So sorry... not a lot of sleep on my end for the last couple of weeks. I understand what you're saying now.

From  what I can see, we keep lat/long in two places:
1. `PostalCodeGeoCoordinate`: which from what I can see, is not used anywhere in the system, which means it should be deleted.
2. `Request`

You're proposing to move lat/long to the `Location` table. If transactionally, the Location data will never change separate from lat/long, then this is a good move.

These are the other models that use Location,. Do they need a lat/long as well?
- Campaign
- Event
- Itinerary
- Organization

If we move lat/long to Location, then these models on writes would need to account for lat/long as well. Not that writing null to lat/long would be "bad", but I just wanted to point out that anything model model using Location could become lat/long aware.

**Comment by stevejgordon** [comment]
@mgmccarthy - No problem at all. :smile: 

So I'm starting to need lat/long on the itinerary to simplify the API calls a bit. That's what triggered my thinking. It feels like if we're storing locations, we make them consistent and store the related lat/long with them. As you say, we would/should ensure that wherever we add a location we also get and store the lat/long from the APIs.

This would be good since we probably will make it easier to cascade a location down from campaign to the event to the itinerary. In those cases the location will already have the lat/long data and we don't have to call the API again.

I'm a bit wary of making addresses related in applications as you can end up in a real mess. Should changing an event address change the itinerary also. I suspect not always. So we do created duplicate addresses inside the locations table I think to prevent that. @MisterJames - I think we touched on this in the past and decided to leave the table there.

In reality addresses are more of a value object and while they may start the same for two domain objects, if one changes, the other might not need to. A bit like a delivery address on an order which should never change, even if the customer address does.

Where we go may depend on how much time we want to put into any changes.

**Comment by mgmccarthy** [comment]
@stevejgordon 

> This would be good since we probably will make it easier to cascade a location down from campaign to the event to the itinerary. In those cases the location will already have the lat/long data and we don't have to call the API again.

agreed about not having to invoke the API multiple times for the same data. Although be careful here, you would need to make sure we don't already have an existing Location of Event/Itinerary in this example before you just took Campaign Location data and wrote it to Event/Itinerary.

> Should changing an event address change the itinerary also. I suspect not always.

agreed.

> In reality addresses are more of a value object

**very much agreed**. Our current implementation is trying to treat `Location` as a value object, [but since EF Core does not have support for `[ComplexType]` yet](https://github.com/aspnet/EntityFramework/issues/246), we have to work around that.

**Comment by stevejgordon** [comment]
@mgmccarthy - Yep 1.2 of EF Core I think would make life better with support from the "ComplexTypes". Not sure if/when spatial data support might be implemented which might also prove useful.

So it feels like the reasonable course of action might be to include an optional lat/long on the location object first and do the work to ensure they are correctly set?
